CHANGELOG.md

```diff
@@ -7,6 +7,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+### Changed
+
+- [breaking-change] Converted all data structures to use the `const generics` MVP
+- [breaking-change] `HistoryBuffer` is now working with const constructors and non-`Copy` data
+- [breaking-change] `HistoryBuffer::as_slice` and others now only return initialized values
+- [breaking-change] `MultiCore`/`SingleCore` is now removed from `spsc::Queue`
+- [breaking-change] `spsc::Queue` is now `usize` only
+- [breaking-change] `spsc::Queue` now sacrifices one element for correctness (see issue #207), i.e. it creates an `N - 1` sized queue instead of the old that generated an size `N` queue
```

Comment:
I think that is counter intuitive. It should create an `N` sized queue and reserve one element more for correctness.

Also depending on the type of element this could waste a lot of space. Maybe some kind of flag would be better for correctness than sacrificing an element.

Reply:
Hi, thanks for the input!

> This implementation detail should not bleed through in my opinion.

I agree that this would be ideal, but unfortunately not possible today.
This is not possible to fix with the current const generics, as math on const generics is not allowed, i.e. `buffer: [MaybeUninit<T>; { N + 1 }]` is not supported.

> Maybe some kind of flag would be better for correctness than sacrificing an element.

This method exists, but is non-trivial to implement. It becomes problematic to have the `Producer` / `Consumer` pairs as they are now memory dependent via this `full` flag.
The `full` flag which is either a `bool` or a counter that checks the fill level of the queue both implementation get a requirement on having CAS operations or `interrupt::free` for accessing and updating this flag.

The new implementation is selected for its simplicity in showing correctness, and correctness is the no. 1 priority.
I however welcome a stab at trying to solve it, I did do some experiments and I was unable getting TSAN happy while keeping simplicity.

> Also depending on the type of element this could waste a lot of space. 

This is, to me, a moot point. If the size of the object is large one should ideally already be using `pool` and send `pool::Box`es.

Reply:
> i.e. `buffer: [MaybeUninit<T>; { N + 1 }]` is not supported.

I though so. That's why I suggested flags.
> This method exists, but is non-trivial to implement.

I feared this would be the case. I just hoped that somebody smarter than me could figure out some way to do it ;)
Now that I think about it, I only ever implemented lock-free queues with `N-1` usable elements and locking queues with a full flag. Never both at the same time, but I thought I was just too lazy to think about it hard enough to figure it out :upside_down_face:.

>correctness is the no. 1 priority.

I totally agree. If it's not guaranteed to be correct it would basically be useless.

> ...one should ideally already be using `pool` and send `pool::Boxes`.

You are probably right about that.

Reply:
Yeah, you know, it's unfortunate.
It will take a bit of getting used to, but this is also what is common for SPSC queues in embedded where almost everyone (including us) base their design on this article: https://www.codeproject.com/Articles/43510/Lock-Free-Single-Producer-Single-Consumer-Circular
A great resource with a lot of details on correctness and pros/cons.

Thanks for you feedback, I really appreciate it!
I hope we could solve it easily but as it is now I think we need to solve it in a future release when const generics have gained more power.